GameServer/Finance/Wallet.cs

```diff
@@ -0,0 +1,138 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DOL.GS.Finance
+{
+    public class Wallet
+    {
+        private readonly Dictionary<Currency, long> balances = new Dictionary<Currency, long>();
+        private GamePlayer owner;
+
+        public Wallet() { }
+
+        public Wallet(GamePlayer owner)
+        {
+            this.owner = owner;
+        }
+
+        public Money GetBalance(Currency currency)
+        {
+            if (currency.IsItemCurrency)
+            {
+                lock (owner.Inventory)
+                {
+                    return currency.Mint(owner.Inventory.GetItemRange(eInventorySlot.FirstBackpack, eInventorySlot.LastBackpack)
+                        .Where(i => i.Template.ClassType.ToLower() == $"Currency.{currency.Name}".ToLower())
+                        .Aggregate(0, (acc, i) => acc + i.Count));
+                }
+            }
+            balances.TryGetValue(currency, out var balance);
+            return currency.Mint(balance);
+        }
+
+        public void AddMoney(Money money)
+        {
+            if (money.Currency.IsItemCurrency)
+            {
+                throw new ArgumentException("You cannot add money of type ItemCurrency.");
+            }
+            lock (balances)
+            {
+                var oldBalance = GetBalance(money.Currency).Amount;
+                var newBalance = oldBalance + money.Amount;
+                SetBalance(money.Currency.Mint(newBalance));
+                SaveToDatabase();
+            }
+        }
+
+        public bool RemoveMoney(Money money)
+        {
+            var currency = money.Currency;
+            if (currency.IsItemCurrency)
+            {
+                lock (owner.Inventory)
+                {
+                    if (GetBalance(currency).Amount < money.Amount) return false;
+
+                    var validCurrencyItemsInventory = owner.Inventory.GetItemRange(eInventorySlot.FirstBackpack, eInventorySlot.LastBackpack)
+                        .Where(i => i.Template.ClassType.ToLower() == $"Currency.{currency.Name}".ToLower());
+                    var remainingDue = (int)money.Amount;
+                    foreach (var currencyItem in validCurrencyItemsInventory)
+                    {
+                        if (currencyItem.Count > remainingDue)
+                        {
+                            owner.Inventory.RemoveCountFromStack(currencyItem, remainingDue);
+                            break;
+                        }
+                        else if (currencyItem.Count == remainingDue)
+                        {
+                            owner.Inventory.RemoveItem(currencyItem);
```

Comment:
After reading your comment about locks, I saw this code, you can just have the RemoveCountFromStack, it will call RemoveItem if the count matches (with a weird `item.AllowAdd = true`...).

Reply:
Good catch!